src/mem/cache/prefetch/indirect_memory_v2.hh

```diff
@@ -0,0 +1,662 @@
+ * (base_address, shift). IMP uses a stream detector to track accesses to array
+ * B. For computing the indirect access parameters, IMP uses a structure called
+ * the IndirectPatternDetector (IPD). After IMP's IPD detects an indirect
+ * pattern, i.e. the once it is able to compute the indirect parameters, IMP
+ * stores the computed pair in its IndirectTable
+ *
+ * For more details about the IMP algorithm, please refer to the paper
+ * mentioned in the references.
+ *
+ * What does this implementation DOESN'T do:
+ * There are some differences between this implementation and the algorithm
+ * described in the paper. Namely, this implementation does not support
+ * indirect prefetching for multi-way and multi-level accesses as described in
+ * section 3.3.2 of the paper
+ */
+
+#ifndef __MEM_CACHE_IMP_HH__
+#define __MEM_CACHE_IMP_HH__
+
+#include <map>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "mem/cache/prefetch/queued.hh"
+#include "mem/packet.hh"
+#include "mem/packet_access.hh"
+#include "params/IMPv2Prefetcher.hh"
+#include "sim/system.hh"
+
+namespace gem5
+{
+    namespace prefetch
+    {
+        namespace IMPv2Internals
+        {
+            template<typename T>
+            void saturating_increment(T& old_value, const T& upper_bound)
+            {
+                if (old_value < upper_bound)
+                    old_value += 1;
+            }
+
+            template<typename T>
+            void saturating_decrement(T& old_value)
+            {
+                if (old_value > 0)
+                    old_value -= 1;
+            }
+
+            /*
+             * The StreamDetector performs two functions:
+             *  1. It detects stream accesses to the indexing array, B
+             *  2. If it is confident about a stream, it returns the predicted
+             *     delta for that stream so that IMP may issue stream
+             *     prefetches using the delta
+             */
+            class StreamDetector
+            {
+                private:
+                    struct StreamTableEntry
+                    {
+                        std::optional<Addr> previous_address;
+                        std::optional<Addr> delta;
+                        uint64_t confidence;
+
+                        StreamTableEntry()
+                        {
+                            previous_address = {};
+                            delta = {};
+                            confidence = 0;
+                        }
+                    };
+
+                    /*
+                     * The confidence threshold above which a stream is
+                     * classified as detected
+                     */
+                    const uint64_t streaming_threshold;
+
+                    /*
+                     * The stream table is indexed using the IP of the load
+                     * instruction
+                     */
+                    std::map<Addr, StreamTableEntry> stream_table;
```

Comment:
nit: ideally this would use the associative cache implementation, as it would allow users to change the replacement and indexing policies as desired. The other prefetchers take this approach, so it would be nice to see it here too. I am not pushing for it, though, as I think this class as a whole should be replaced by the stride prefetcher (either in this PR, or in the future).

Reply:
The prefetch table (which is divided into two distinct parts, the `stream_table` and the `indirect_table`) is a direct mapped structure, which is indexed by the IP, as per the paper's description (see Figure 5 of the [paper](https://pages.cs.wisc.edu/~yxy/pubs/imp.pdf)). Since this structure is a direct mapped structure, converting it into an associative cache would lead to more confusion. So, I think we should leave it as it is. What do you think?

Reply:
I tend to think that since gem5 is used by researchers, we want to make their lives easier in case they want to expand existing state-of-the-art algorithms. As such, I think that using a generic cache will be better, even if the original paper uses a specific configuration. That said, as I said this request is is a nitpick, and can be ignored without a problem.